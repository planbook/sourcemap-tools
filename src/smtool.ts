import type { StackFrame } from 'stacktrace-parser'

import {
  type NullableMappedPosition,
  type RawIndexMap,
  type RawSourceMap,
  SourceMapConsumer,
} from 'source-map'
import * as stackTraceParser from 'stacktrace-parser'

export async function transform(
  stackTrace: string,
  maps: Record<string, RawIndexMap | RawSourceMap | string>,
) {
  const { frames, message } = parseStackTrace(stackTrace)

  const newStack = [message]
  const transformed = await Promise.all(frames.map(sf => processStackFrame(sf, maps)))
  const result = newStack.concat(transformed).join('\n')
  return result
}

export type ParsedStackTrace = {
  frames: StackFrame[]
  message: string
}

export function parseStackTrace(stackTrace: string): ParsedStackTrace {
  const frames = stackTraceParser.parse(stackTrace)
  const message = extractErrorMessage(stackTrace)

  return { frames, message }
}

function extractErrorMessage(stackTrace: string) {
  return stackTrace.split('\n')[0]
}

async function processStackFrame(
  stackFrame: StackFrame,
  maps: Record<string, RawIndexMap | RawSourceMap | string>,
) {
  const originalPosition = await tryGetOriginalPosition(stackFrame, maps)

  return originalPosition
    ? generateStackTraceLine(getPosition(originalPosition))
    : generateStackTraceLine(getPosition(stackFrame))
}

async function tryGetOriginalPosition(
  stackFrame: StackFrame,
  maps: Record<string, RawIndexMap | RawSourceMap | string>,
): Promise<NullableMappedPosition | null> {
  let result: NullableMappedPosition | null = null

  const { column, file, line } = getPosition(stackFrame)

  if (!file || !maps[file] || !line || !column) {
    return null
  }

  await SourceMapConsumer.with(maps[file], null, consumer => {
    result = consumer.originalPositionFor({ column, line })
  })

  return result
}

type Position = {
  column: number
  file: string
  line: number
  method: string
}

function generateStackTraceLine(position: Position) {
  const { column, file, line, method } = position
  return `  at ${method || '<unknown>'} (${file}:${line}:${column})`
}

function getPosition(position: NullableMappedPosition | StackFrame): Position {
  let method = ''
  let file = ''
  let line = -1
  let column = -1

  if (isStackFrame(position)) {
    method = position.methodName
    file = position.file || ''
    line = position.lineNumber || -1
    column = position.column || -1
  }

  if (isNullableMappedPosition(position)) {
    method = position.name || ''
    file = position.source || ''
    line = position.line || -1
    column = position.column || -1
  }

  return { column, file, line, method }
}

function isStackFrame(position: NullableMappedPosition | StackFrame): position is StackFrame {
  return 'lineNumber' in position
}

function isNullableMappedPosition(
  position: NullableMappedPosition | StackFrame,
): position is NullableMappedPosition {
  return 'source' in position
}

export function createConsumer(rawSourceMap: string) {
  return new SourceMapConsumer(rawSourceMap)
}
