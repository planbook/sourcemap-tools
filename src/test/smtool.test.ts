import { transform } from '../smtool.ts'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'

import stackTrace from './mocks/stacktrace'
import { expect } from 'vitest'

const indexMap = readFileSync(
  resolve(__dirname, 'mocks/sourcemaps/index-d803759c.js.map'),
).toString()
const vendorMap = readFileSync(
  resolve(__dirname, 'mocks/sourcemaps/vendor-221d27ba.js.map'),
).toString()

describe('transform', () => {
  test('transforms', async () => {
    expect(
      await transform(stackTrace, {
        'index-d803759c.js': indexMap,
        'vendor-221d27ba.js': vendorMap,
      }),
    ).toMatchSnapshot()
  })
})
