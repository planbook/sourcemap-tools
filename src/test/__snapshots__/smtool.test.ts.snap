// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > transforms 1`] = `
"Uncaught Error: Error!
  at <unknown> (../../src/utils.ts:2:8)
  at throwError (../../src/App.tsx:5:15)
  at <unknown> (../../node_modules/react-dom/cjs/react-dom.production.min.js:54:316)
  at <unknown> (../../node_modules/react-dom/cjs/react-dom.production.min.js:54:470)
  at <unknown> (../../node_modules/react-dom/cjs/react-dom.production.min.js:55:34)
  at Ub (../../node_modules/react-dom/cjs/react-dom.production.min.js:105:67)
  at nf (../../node_modules/react-dom/cjs/react-dom.production.min.js:106:379)
  at se (../../node_modules/react-dom/cjs/react-dom.production.min.js:117:103)
  at a (../../node_modules/react-dom/cjs/react-dom.production.min.js:274:41)
  at Gb (../../node_modules/react-dom/cjs/react-dom.production.min.js:52:374)"
`;
